p2p/raintree/network.go

```diff
@@ -25,46 +25,71 @@ type rainTreeNetwork struct {
 
 	peersManager *peersManager
 
+	cleanupLayerEnabled    bool // debug config only
+	redundancyLayerEnabled bool
+
 	// TECHDEBT(drewsky): What should we use for de-duping messages within P2P?
 	mempool map[uint64]struct{} // TODO (drewsky) replace map implementation (can grow unbounded)
 }
 
-func NewRainTreeNetwork(addr cryptoPocket.Address, addrBook typesP2P.AddrBook) typesP2P.Network {
+func NewRainTreeNetwork(addr cryptoPocket.Address, addrBook typesP2P.AddrBook, cfg modules.P2PConfig) typesP2P.Network {
 	pm, err := newPeersManager(addr, addrBook)
 	if err != nil {
 		log.Println("[ERROR] Error initializing rainTreeNetwork peersManager: ", err)
 	}
 
 	n := &rainTreeNetwork{
-		selfAddr:     addr,
-		peersManager: pm,
-		mempool:      make(map[uint64]struct{}),
+		selfAddr:               addr,
+		peersManager:           pm,
+		redundancyLayerEnabled: cfg.GetUseRainTreeRedundancyLayer(),
+		cleanupLayerEnabled:    cfg.GetUseRainTreeCleanupLayer(),
+		mempool:                make(map[uint64]struct{}),
 	}
 
 	return typesP2P.Network(n)
 }
 
 func (n *rainTreeNetwork) NetworkBroadcast(data []byte) error {
-	return n.networkBroadcastAtLevel(data, n.peersManager.getNetworkView().maxNumLevels, getNonce())
+	return n.networkBroadcastAtLevel(data, int32(n.peersManager.getNetworkView().maxNumLevels), getNonce())
 }
 
-func (n *rainTreeNetwork) networkBroadcastAtLevel(data []byte, level uint32, nonce uint64) error {
-	// This is handled either by the cleanup layer or redundancy layer
-	if level == 0 {
-		return nil
-	}
+func (n *rainTreeNetwork) networkBroadcastAtLevel(data []byte, level int32, nonce uint64) error {
 	msg := &typesP2P.RainTreeMessage{
 		Level: level,
 		Data:  data,
 		Nonce: nonce,
 	}
+
+	// Redundancy layer
+	if level == 0 {
+		if n.redundancyLayerEnabled {
+			level, msg.Level = n.redundancyLayer()
+		} else {
+			if err := n.demote(msg); err != nil {
+				log.Println("Error demoting self during RainTree message propagation: ", err)
+			}
+			return nil
+		}
+	}
+
+	targets := n.getTargetsAtLevel(uint32(level))
+
+	// Cleanup layer
+	if level == -1 {
+		if !n.cleanupLayerEnabled {
+			return nil
+		}
+		targets = n.cleanupLayer()
```

Comment:
<img width="901" alt="Screenshot 2022-12-01 at 7 36 47 PM" src="https://user-images.githubusercontent.com/1892194/205209364-233fe9b4-5397-4030-adcd-da08830acdcd.png">

Redundancy is like UDP, but at the application layer.
Cleanup is like TCP, but at the application layer.

Out of scope. Let's add a TODO.

Reply:
If a basic implementation (happy paths) of redundancy and cleanup layer is not in scope, what is in scope then? 
Can we clarify that?

From my understanding, since this PR was in review and you guys mentioned that it was "very close" I am definitely missing something................. or maybe this is just an @Olshansk joke and I am the victim.... where's the camera? :) 

Seriously: **Correct me if I am mistaken.**

From the code I see that what was added was:
- feature flags
- tests calls with feature flags on/off
- some weird handling of the `level` 
- redundancy and clean-up essentially implemented to just spit out targets

I can't reconcile the code with the paper so which one of the two is wrong? 🤷‍♂️